File: answers/One Edit/Python/one_edit.py
# Started with this error:
# Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

# Answer: 
# 1. Open a command prompt (or terminal down below)
# 2. 'where python'
# 3. Ended up just restarting computer lol thanks to this stack overflow user: 
### https://stackoverflow.com/questions/52332554/vscode-the-term-python-is-not-recognized-but-py-works
# 4. In the bottom right, I had not selected an interpreter
# - You can either choose your Anaconda interpreter (recommended), or the python on your system that you found with 'where python'
##### ie. C:\Users\Myles\anaconda3\python.exe
##### vs. C:\Users\Myles\AppData\Local\Programs\Python

# Bonus: go to your search bar and search for Manage App Execution Aliases. 
# You will find the attached screen and you need to turn off App Installers


# Make a class with a basic name ie. Program
class Program:
    def __init__(self, stringOne, stringTwo):
        self.stringOne = stringOne
        self.stringTwo = stringTwo

    def oneEdit(self):
        # Write your code here.
        #stringOne = self.stringOne
        #stringTwo = self.stringTwo
        print(f"Running oneEdit() from class 'Program' with the following attributes:\nself.stringOne = {self.stringOne} \nself.stringTwo = {self.stringTwo}")
        
        # Get lengths
        len1 = len(self.stringOne)
        len2 = len(self.stringTwo)
        if len1 < 2 and len2 < 2:
            return True
            
        len_diff = abs(len1 - len2)
        if len_diff > 1:
            return False

        i = 0
        j = 0
        counter = 0
        if len_diff == 1:
            # Iterate through - if you find a difference, move ONLY the bigger string ahead 1
            while i < len1 and j < len2:
                if self.stringOne[i] != self.stringTwo[j]:
                    counter += 1
                    if len1 > len2:
                        i += 1
                    elif len1 < len2:
                        j += 1
                elif self.stringOne[i] == self.stringTwo[j]:
                    i += 1
                    j += 1
            
        elif len_diff == 0:
            # Iterate through - if you find a difference, counter += 1
            while i < len1:
                if self.stringOne[i] != self.stringTwo[i]:
                    counter += 1
                i += 1

        # Find how many iterations to do (equal to the length of the smaller string)
        # Return depending on counter
        return_val = True if counter <= 1 else False
        print(f"return_val: {return_val}")
        return return_val


# We have created the following
### - 'Program' Class
### - 'init' method
### - 'one_edit' method

# We can now make an instance of Program and get our results
instance_of_program = Program("myles", "jamison")
print(instance_of_program.oneEdit())

# Edit values, do it again
print("\nRunning the method again, this time with new values!")
instance_of_program.stringOne = "Max"
instance_of_program.stringTwo = "Maxy"
print(instance_of_program.oneEdit())

File: answers/One Edit/Java/one_edit.java
// Imports
//import java.util.*; it doesn't get used so cancelled it out 

// Creating a class named 'one_edit'
// Was getting error 'public type one_edit must be defined in its own file' 
// Until I changed 'public class one_edit' into 'class one_edit'
//public 
class one_edit { // Make sure that the class name is the same as the file (ie. one_edit.java)
    public String stringOne;
    public String stringTwo;

    // Constructor 
    public one_edit(String s1, String s2) {
        stringOne = s1;
        stringTwo = s2;
        // either of these can work!
        //this.stringOne = s1
        //this.stringTwo = s2
    }

    // Method
    public boolean oneEdit() {
        // Write your code here.
        String s = String.format("Running method oneEdit from an object from class 'one_edit' with the following:\nstringOne = %s \nstringTwo = %s \n !", stringOne, stringTwo);
        System.out.println(s);

        // Get lengths
        int len1 = stringOne.length();
        int len2 = stringTwo.length();

        if (Math.abs(len1 - len2) > 1) {
            System.out.println("The difference between stringOne and stringTwo is > 1: return false");
            return false;
          }

        boolean madeEdit = false;
        int i = 0;
        int j = 0;

        while (i < len1 && j < len2) {
            if (stringOne.charAt(i) != stringTwo.charAt(j)) {
                if (madeEdit) {
                  System.out.println("Making a 2nd edit: return false");
                  return false;
                }

                else if (!madeEdit) {
                  System.out.println("Making first edit... next one means false");
                  madeEdit = true;
        
                  if (len1 > len2) {
                    i++;
                  }

                  else if (len1 < len2) {
                    j++;
                  }

                  else {
                    i++;
                    j++;
                  }
                  
                }
              }
        
              else if (stringOne.charAt(i) == stringTwo.charAt(j)) {i++; j++;}

        } // End of while loop

        // Return the answer
        s = String.format("Made it to the end, %s and %s are good to go! \n", stringOne, stringTwo);
        System.out.println(s);
        return true;

    } // End of 'oneEdit' method declaration

    // Declare the main
    public static void main(String[] args) {
        // Create an instance of our class 'one_edit'
        
        // Note: As long as we turn this into an object, we don't need 'static' above 
        // ie. public static boolean vs. public boolean
        ///// one_edit myInstance = new one_edit();
        ///// System.out.println(myInstance.instanceMethod()); //will work
        one_edit instance1 = new one_edit("grab", "grabs"); //System.out.println();

        // Run the 'oneEdit()' method with instance1
        boolean bools;
        bools = instance1.oneEdit();
        System.out.println(bools);


        // Edit the values of instance1
        instance1.stringOne = "abcd";
        instance1.stringTwo = "abcdefg";

        // Run the method one more time!
        bools = instance1.oneEdit();
        System.out.println(bools);



        // return ??? I don't think we need this in Java



    } // End of public static void main()


    
} // End of class one_edit declaration
    



